### PR TITLE
Update screens for deposit and withdrawal flows

### PR DESCRIPTION
**Pull request by kkosiorowska**
Ref https://github.com/thesis/acre/issues/811
Ref https://github.com/thesis/acre/issues/277
Closes https://github.com/thesis/acre/issues/814
Closes https://github.com/thesis/acre/issues/813

This PR updates screens for deposit and withdrawal flows. The "Skeleton" loading screen from both Deposit and Withdraw flows is no longer needed. The new flow adds two steps:

- "Opening your wallet for signature" Screen
- "Awaiting Transaction" Screen

Flow:
1. User initiates deposit: After pressing the Deposit/Withdraw CTAs, the user first sees the “Opening your wallet for signature” screen 
2. Display wallet modal
3. Transition to ‘Awaiting Transaction’: Once the wallet modal is prompted, switch to the “Awaiting Transaction” screen while the user completes the signature.
4. Return to dapp: After signing with the wallet, the user returns to your dapp, where the “Awaiting Transaction” screen remains visible as needed.
5. Confirm deposit/withdraw: Once transaction is complete, display the “Deposit received” or "Withdrawal inititated" screens to confirm completion.

https://github.com/user-attachments/assets/899263a7-0895-48ba-881e-c61b70e4c32d

### CODE REVIEW

File: dapp/src/components/TransactionModal/WalletInteractionModal.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    progressProps: { transform: "scaleX(-1)" },
+  },
+}
+
+export default function WalletInteractionModal({
+  step,
+}: {
+  step: WalletInteractionStep
+}) {
+  const type = useActionFlowType()
```
Comment (r-czajkowski) [inline dapp/src/components/TransactionModal/WalletInteractionModal.tsx:50]:
```suggestion
  const action = useActionFlowType()
```

or
```suggestion
  const actionType = useActionFlowType()
```

File: sdk/src/lib/redeemer-proxy.ts [hunk 2]
```diff
@@ -25,6 +26,8 @@ export default class OrangeKitTbtcRedeemerProxy implements TbtcRedeemerProxy {
 
   #sharesAmount: bigint
 
+  #builtDataStepCallback?: BuiltDataStepCallback
```
Comment (r-czajkowski) [inline sdk/src/lib/redeemer-proxy.ts:29]: Can't we use `onSignMessageStepCallback`?

Comment (r-czajkowski) [inline sdk/src/lib/redeemer-proxy.ts:29]: Maybe we should rename to 🤔 :
```suggestion
  #buildDataStepCallback?: BuiltDataStepCallback
```

or

```suggestion
  #dataBuiltStepCallback?: BuiltDataStepCallback
```

File: sdk/src/lib/redeemer-proxy.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { OrangeKitSdk } from "@orangekit/sdk"
 import { AcreContracts } from "./contracts"
 import { BitcoinProvider } from "./bitcoin"
 
+export type BuiltDataStepCallback = (safeTxData: Hex) => Promise<void>
```
Comment (r-czajkowski) [inline sdk/src/lib/redeemer-proxy.ts:10]: Same as above

File: dapp/src/components/TransactionModal/WalletInteractionModal.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        <HStack minW={80} spacing={5}>
+          <AcreSignIcon {...ICON_STYLES} />
+          {/* TODO: Create correct progress bar */}
+          <Progress
+            size="sm"
+            bg="gold.200"
+            isIndeterminate
+            {...progressProps}
+          />
+          <Image src={connector?.icon} p={2} bg="black" {...ICON_STYLES} />
```
Comment (r-czajkowski) [inline dapp/src/components/TransactionModal/WalletInteractionModal.tsx:70]: Let's add `alt="Connector icon`.

Comment (kkosiorowska) [reply dapp/src/components/TransactionModal/WalletInteractionModal.tsx]: https://github.com/thesis/acre/pull/821/commits/f3f18bec965b9656ade4fb3b65362186c96c1902

File: sdk/src/lib/redeemer-proxy.ts (hunk 2 above)
Comment (kkosiorowska) [reply sdk/src/lib/redeemer-proxy.ts:29]: Unfortunately, no. Withdrawal flow is the following three steps:

1. Preparing transaction
2. Opening wallet for signature
3. Awaiting signature confirmation

So we need to know when the data for the transaction is ready to display step 2. The status change in `onSignMessageStepCallback` matches the change from step 2 to step 3.

File: dapp/src/components/TransactionModal/WalletInteractionModal.tsx (hunk 1 above)
Comment (kkosiorowska) [reply dapp/src/components/TransactionModal/WalletInteractionModal.tsx]: https://github.com/thesis/acre/pull/821/commits/3fbfbef368e4d1a07c9e922e0865354b3d4e47d0

File: sdk/src/lib/redeemer-proxy.ts (hunk 2 above)
Comment (kkosiorowska) [reply sdk/src/lib/redeemer-proxy.ts]: https://github.com/thesis/acre/pull/821/commits/8ceb6b4e73b454b77008971ce316228eaafffea2

File: sdk/src/lib/redeemer-proxy.ts (hunk 3 above)
Comment (kkosiorowska) [reply sdk/src/lib/redeemer-proxy.ts]: https://github.com/thesis/acre/pull/821/commits/8ceb6b4e73b454b77008971ce316228eaafffea2

File: sdk/src/lib/redeemer-proxy.ts (hunk 2 above)
Comment (kkosiorowska) [reply sdk/src/lib/redeemer-proxy.ts]: > Isn't the data ready when we ask for a signature?

Yeah the data is ready at that time.

> Can't we display the preparing transaction step just after invoking the withdrawal function, then change the status in onSignMessageStepCallback to Opening wallet for signature and in messageSignedStepCallback display Awaiting signature confirmation?

The`Opening wallet for signature` step should change to `Awaiting signature confirmation` when the wallet window opens. Therefore, we need the additional state of `dataBuiltStepCallback`. The status change in `onSignMessageStepCallback` indicates that the wallet should be open.

File: dapp/src/hooks/useDepositBTCTransaction.ts [hunk 5]
```diff
@@ -24,6 +25,7 @@ export function useDepositBTCTransaction(
   const [transactionHash, setTransactionHash] = useState<string | undefined>(
     undefined,
   )
+  const [inProgress, setInProgress] = useState<boolean>(false)
```
[the remark this answers is not in the document]
Comment (kkosiorowska) [reply dapp/src/hooks/useDepositBTCTransaction.ts]: https://github.com/thesis/acre/pull/821/commits/1717ea5b817f7f1ba6158f30975579a128027b13

File: dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx [hunk 6]
```diff
@@ -62,10 +60,8 @@ export default function DepositBTCModal() {
     [onError, handlePause],
   )
 
-  const { sendBitcoinTransaction, transactionHash } = useDepositBTCTransaction(
-    onDepositBTCSuccess,
-    onDepositBTCError,
-  )
+  const { sendBitcoinTransaction, transactionHash, inProgress } =
```
[the remark this answers is not in the document]
Comment (kkosiorowska) [reply dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx]: https://github.com/thesis/acre/pull/821/commits/1717ea5b817f7f1ba6158f30975579a128027b13

File: dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx [hunk 7]
```diff
@@ -36,81 +34,68 @@ export default function DepositBTCModal() {
   }, [dispatch, handleBitcoinBalanceInvalidation])
 
   const onError = useCallback(
-    (error?: unknown) => {
-      console.error(error)
-      dispatch(setStatus(PROCESS_STATUSES.FAILED))
-    },
+    () => dispatch(setStatus(PROCESS_STATUSES.FAILED)),
```
Comment (r-czajkowski) [inline dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx:37]: In my opinion, we should log errors. It always adds some context. We can improve it eg. `console.error("Bitcoin deposit failed", error)`

File: dapp/src/hooks/useDepositBTCTransaction.ts [hunk 8]
```diff
@@ -1,65 +1,51 @@
...
-  transactionHash?: string
-  sendBitcoinTransaction: (
-    ...params: SendBitcoinTransactionParams
-  ) => Promise<void>
-}
-
 export function useDepositBTCTransaction(
-  onSuccess?: OnSuccessCallback,
+  onSuccess?: OnSuccessCallback<string>,
   onError?: OnErrorCallback,
```
Comment (r-czajkowski) [inline dapp/src/hooks/useDepositBTCTransaction.ts:8]: We can pass `onSuccess` and. `onError` to `mutate` function.

File: dapp/src/hooks/useExecuteFunction.ts
Comment (r-czajkowski) [inline dapp/src/hooks/useExecuteFunction.ts:1]: Looks like we can use the `useMutation` hook here as well. Ofc non-blocking comment, just flagging.

File: dapp/src/hooks/useDepositBTCTransaction.ts (hunk 8 above)
Comment (kkosiorowska) [reply dapp/src/hooks/useDepositBTCTransaction.ts]: https://github.com/thesis/acre/pull/821/commits/26da7a55c84e4bbd6f27156a5208ac63b04c5429

File: dapp/src/hooks/useExecuteFunction.ts
Comment (kkosiorowska) [reply dapp/src/hooks/useExecuteFunction.ts:1]: https://github.com/thesis/acre/pull/821/commits/26da7a55c84e4bbd6f27156a5208ac63b04c5429

File: dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx [hunk 9]
```diff
@@ -62,55 +70,68 @@ export default function DepositBTCModal() {
...
+      const satoshis = Number(amount)
+
+      // @ts-expect-error adjust types to handle bitcoin wallet wrappers
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-assignment
+      const txhash: string = await client.sendBitcoin(recipient, satoshis)
+      return txhash
+    },
+    onSuccess: onDepositBTCSuccess,
+    onError: onDepositBTCError,
+  })
```
Comment (r-czajkowski) [inline dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx:101]: We could keep this code in `useDepositBTCTransaction`.

**Review by r-czajkowski — DISMISSED** [review]

File: dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx (hunk 9 above)
Comment (kkosiorowska) [reply dapp/src/components/TransactionModal/ActiveStakingStep/DepositBTCModal.tsx]: https://github.com/thesis/acre/pull/821/commits/1440e0796c9e8561e2fb8b22ac3a0c482670cd18

**Review by r-czajkowski — APPROVED** [review]

**Pull request merged**